pytensor/tensor/pad.py

```diff
@@ -0,0 +1,536 @@
+) -> TensorVariable:
+    padded, area_slice, pad_width = _symbolic_pad(x, pad_width)
+    end_values = as_tensor(end_values)
+    end_values = broadcast_to(end_values, as_tensor((padded.ndim, 2)))
+
+    for axis in range(padded.ndim):
+        width_pair = pad_width[axis]
+        end_value_pair = end_values[axis]
+        edge_pair = _get_edges(padded, axis, width_pair)
+        dim_shape = padded.shape[axis]
+        left_slice, right_slice = _get_padding_slices(dim_shape, width_pair, axis)
+
+        left_ramp, right_ramp = (
+            linspace(
+                start=end_value,
+                stop=specify_broadcastable(edge, axis).squeeze(axis),
+                num=width,
+                endpoint=False,
+                dtype=padded.dtype,
+                axis=axis,
+            )
+            for end_value, edge, width in zip(end_value_pair, edge_pair, width_pair)
+        )
+
+        # Reverse the direction of the ramp for the "right" side
+        right_ramp = right_ramp[_slice_at_axis(slice(None, None, -1), axis)]  # type: ignore
+
+        padded = set_subtensor(padded[left_slice], left_ramp)
+        padded = set_subtensor(padded[right_slice], right_ramp)
+
+    return padded
+
+
+def flip(x, axis=None):
+    if axis is None:
+        index = ((slice(None, None, -1)),) * x.ndim
+    else:
+        if isinstance(axis, int):
+            axis = [axis]
+        index = [
+            slice(None, None, -1) if i in axis else slice(None, None, None)
+            for i in range(x.ndim)
+        ]
+    return x[index]
+
+
+def _wrap_pad(x: TensorVariable, pad_width: TensorVariable) -> TensorVariable:
+    pad_width = broadcast_to(pad_width, as_tensor((x.ndim, 2)))
+
+    for axis in range(x.ndim):
+        size = x.shape[axis]
+
+        # Compute how many complete copies of the input will be padded on this dimension, along with the amount of
+        # overflow on the final copy
+        repeats, (left_remainder, right_remainder) = pt_divmod(pad_width[axis], size)
+
+        # In the next step we will generate extra copies of the input, and then trim them down to the correct size.
+        left_trim = size - left_remainder
+        right_trim = size - right_remainder
+
+        # The total number of copies needed is always the sum of the number of complete copies to add, plus the original
+        # input itself, plus the two edge copies that will be trimmed down.
+        total_repeats = repeats.sum() + 3
+
+        # Create a batch dimension and clone the input the required number of times
+        parts = expand_dims(x, (0,)).repeat(total_repeats, axis=0)
+
+        # Move the batch dimension to the active dimension
+        parts = moveaxis(parts, 0, axis)
+
+        # Ravel the active dimension while preserving the shapes of the inactive dimensions. This will expand the
+        # active dimension to have the correctly padded shape, plus excess to be trimmed
```

Comment:
Make this a helper in shape.py? A bunch of rewrites and helper functions need this "flatten_axes”

Reply:
I'm not against it but I'm also not clear what the signature should be. You need two arrays, `arr1` and `arr2` with `arr1.ndim > arr2.ndim` where the shapes of `arr2` are in the set of the shapes of `arr1`. Then we can flatten the additional dimensions of `arr1` according to the shapes of `arr2`.

It's quite specific. Are you sure that's something that will come up enough to merit a helper?

Reply:
I meant something more restrict, you have one arr and you flatten n dims and place them at a given axis?

Reply:
But I haven't given it much thought, so if it doesn't sound obviously useful let's just move on